### ISSUE TITLE
[patch] generalize command line option -D

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#729]
-D <setting=value> works only if the value is either int or string.
It would be handy for me to pass -D'exclude_patterns=[]', so I propose to modify slightly the code:

```
#!diff

--- a/sphinx/cmdline.py Sat Jul 09 12:08:18 2011 +0200
+++ b/sphinx/cmdline.py Tue Jul 12 14:43:14 2011 +0200
@@ -128,14 +128,14 @@
             confdir = None
         elif opt == '-D':
             try:
-                key, val = val.split('=')
+                key, val = val.split('=', 1)
             except ValueError:
                 print >>sys.stderr, ('Error: -D option argument must be '
                                      'in the form name=value.')
                 return 1
             try:
-                val = int(val)
-            except ValueError:
+                val = eval(val)
+            except:
                 if likely_encoding:
                     try:
                         val = val.decode(likely_encoding)
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/729
- Originally reported by: Marcin Wojdyr
- Originally created at: 2011-07-12T14:55:55.815

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2011-09-22 10:14:14+00:00_

I can see this has merits, but I don't like the eval here. Do you think another way could be found to allow list arguments?

**Comment by shimizukawa** [comment]
_From Marcin Wojdyr on 2011-09-22 21:30:13+00:00_

[[http://docs.python.org/library/ast.html?highlight=literal#ast.literal_eval|ast.literal_eval()]] could be used instead.

**Comment by shimizukawa** [comment]
_From Robert Lehmann on 2011-09-28 12:09:48+00:00_

What's our backwards compatibility policy?  With that change applied, {{{-D project=None}}} (whatever merits that particular value does or does not have) would suddenly break.  Other surprising effects include silently swallowing surrounding quotes and misparsing version numbers.  Plus {{{ast.literal_eval}}} is new in **Python 2.6**.

I am unsure how complex this mechanism needs to grow but the only identified use case so far is to **reset** a configuration option, right?  If that's the case and no other user story comes up, I'd propose having {{{-D name}}} (without a {{{=value}}}) reset an option.  This probably needs to be aware of the type of a configuration value to come up with the correct null object (empy string/empty list).  For reasons of non-suprisingness, another option-- say {{{-R}}} --could be added for that purpose too.

I'd also like to note we should not make the command-line interface //much// more complicated than it is, because the OP's use case is fully satisfied with using {{{-c}}} with an alternative {{{conf.py}}}.

**Comment by shimizukawa** [comment]
_From Anonymous on 2012-01-25 14:06:31+00:00_

I wanted to use -D today to configure all settings (thus also passing -C). It currently cannot be done, because complex data structures cannot be passed (for example latex_documents)

So my first stab at a solution was to try and pass custom variable names with -D in and then using those variables from within the conf.py. This fails because the settings that are set with -D is not passed into conf.py, but rather overrides their values after the fact. Which is kind of the point, and obvious when I came to think about it.

The proposed change in this bug report would probably help me solve my problem, but I'm not sure if it is an elegant solution.

**Comment by shimizukawa** [comment]
_From [holmboe](https://bitbucket.org/holmboe) on 2012-01-25 14:08:21+00:00_

The above comment was written by me, just wasn't logged in. For future reference.

**Comment by shimizukawa** [comment]
_From Marcin Wojdyr on 2012-01-25 22:27:02+00:00_

[holmboe](https://bitbucket.org/holmboe): a workaround is to pass argument to conf.py through environement variable:

{{{FOO=1 sphinx-build ...}}}

and then check it in conf.py using {{{os.getenv('FOO')}}}

**Comment by shimizukawa** [comment]
_From [holmboe](https://bitbucket.org/holmboe) on 2012-01-26 08:14:44+00:00_

[Marcin Wojdyr](https://bitbucket.org/wojdyr) That is true, didn't think of that. The solution I came to was to write a pickle to file and then read it at the start of conf.py. I'm not entirely sure that I'm happy with my solution, but it works and I get access to all the data I need inside conf.py. Thanks for your suggestion.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-01-19 17:14:24+00:00_

This has now been implemented for list types as well in d98e70c13eff.
